src/tzkey.rs

```diff
@@ -19,20 +19,28 @@ pub fn jwk_from_tezos_key(tz_pk: &str) -> Result<JWK, Error> {
             }),
         ),
         "edsk" => {
-            let mut sk_bytes = bs58::decode(&tz_pk).with_check(None).into_vec()?[4..].to_owned();
-            if sk_bytes.len() == 64 {
-                sk_bytes = sk_bytes[..32].to_vec();
-            }
+            let sk_bytes = bs58::decode(&tz_pk).with_check(None).into_vec()?[4..].to_owned();
             let pk_bytes;
             #[cfg(feature = "ring")]
             {
                 use ring::signature::KeyPair;
-                let keypair = ring::signature::Ed25519KeyPair::from_seed_unchecked(&sk_bytes)?;
+                let keypair = if sk_bytes.len() == 64 {
+                    ring::signature::Ed25519KeyPair::from_seed_and_public_key(
+                        &sk_bytes[..32],
+                        &sk_bytes[32..],
+                    )?
+                } else {
+                    ring::signature::Ed25519KeyPair::from_seed_unchecked(&sk_bytes)?
+                };
                 pk_bytes = keypair.public_key().as_ref().to_vec()
             }
             #[cfg(feature = "ed25519-dalek")]
             {
-                let sk = ed25519_dalek::SecretKey::from_bytes(&sk_bytes)?;
+                let sk = ed25519_dalek::SecretKey::from_bytes(if sk_bytes.len() == 64 {
+                    &sk_bytes[..32]
+                } else {
+                    &sk_bytes
+                })?;
```

Comment:
For length == 64, `ed25519_dalek::Keypair::from_bytes(&sk_bytes)` could be used so that the dalek library can check the consistency of the keypair. Then you could get the SecretKey and PublicKey from the Keypair struct.

Reply:
https://docs.rs/ed25519-dalek/1.0.1/ed25519_dalek/struct.Keypair.html#warning It doesn't actually :(

Reply:
Oh, ok :sweat_smile: